stumpy/Mmotifs.py

```diff
@@ -0,0 +1,297 @@
+
+    atol : float, default 1e-8
+        The absolute tolerance parameter. This value will be added to `max_distance`
+        when comparing distances between subsequences.
+
+    max_motifs: int, default 1
+        The maximum number of motifs to return
+
+    cutoff: numpy.ndarray or float, default None
+    The largest matrix profile value (distance) for each dimension of the
+    multidimensional matrix profile that a multidimenisonal candidate motif is
+    allowed to have.
+    If cutoff is only one value, these value will be applied to every dimension.
+
+    visualize_mdl: boolean, default 'False'
+        Visualize MDL results
+
+    Returns
+    -------
+    motif_matches_distances: numpy.ndarray
+        The distances corresponding to a set of subsequence matches for each motif.
+
+    motif_matches_indices: numpy.ndarray
+        The indices corresponding to a set of subsequences matches for each motif.
+
+    subspace: numpy.ndarray
+        A numpy.ndarray consisting of arrays that contain the `k`-dimensional
+        subspace for each motif.
+    """
+    if max_motifs < 1:  # pragma: no cover
+        logger.warning(
+            "The maximum number of motifs, `max_motifs`, "
+            "must be greater than or equal to 1"
+        )
+        logger.warning("`max_motifs` has been set to `1`")
+        max_motifs = 1
+
+    # Calculate subsequence length
+    T = _preprocess(T)  # convert dataframe to array if necessary
+    m = T.shape[-1] - P.shape[-1] + 1
+
+    # Calculate exclusion zone
+    excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
+
+    if max_matches is None:
+        max_matches = np.inf
+
+    motif_matches_distances = []
+    motif_matches_indices = []
+    subspace = []
+
+    # Identify the first multidimensional motif as candidate motif and get its
+    # nearest neighbor
+    candidate_idx = np.argsort(P, axis=1)[:, 0]
+    nn_idx = I[np.arange(len(candidate_idx)), candidate_idx]
+
+    # Iterate over each motif
+    while len(motif_matches_distances) < max_motifs:
+
+        # Choose dimension k using MDL
+        mdl, subspaces = stumpy.mdl(T, m, candidate_idx, nn_idx)
+        k = np.argmin(mdl)
+
+        # Calculate default cutoff value
+        if cutoff is None:
+            P_copy = P.copy().astype(np.float64)
+            P_copy[np.isinf(P_copy)] = np.nan
+            P_copy = P_copy[k, :]
+            cutoff_var = np.nanmax(
+                [np.nanmean(P_copy - 2.0 * np.nanstd(P_copy)), np.nanmin(P_copy)]
+            )
+            cutoff = None
+        else:
+            if isinstance(cutoff, np.ndarray) or isinstance(cutoff, list):
+                cutoff_var = cutoff[k]
+            else:
+                cutoff_var = cutoff
+
+        # Get k-dimensional motif
+        motif_idx = candidate_idx[k]
+        motif_value = P[k, motif_idx]
+        if motif_value > cutoff_var or not np.isfinite(motif_value):
+            break
+
+        if visualize_mdl:

```

Comment:
Please remove this part and consider returning both a list of `motif_mdls` and `motif_subspaces` back to the user where necessary